stdlib/public/core/StringWordBreaking.swift

```diff
@@ -0,0 +1,690 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension _StringGuts {
+  @inline(__always)
+  internal func roundDownToNearestWord(
+    _ i: String._WordView.Index
+  ) -> String._WordView.Index {
+    _internalInvariant(i._encodedOffset <= count)
+
+    let offset = i._encodedOffset
+
+    if offset == 0 || offset == count {
+      return i
+    }
+
+    return _slowRoundDownToNearestWord(i)
+  }
+
+  @inline(never)
+  internal func _slowRoundDownToNearestWord(
+    _ i: String._WordView.Index
+  ) -> String._WordView.Index {
+    let offset = i._encodedOffset
+    let start = _opaquePreviousWordIndex(endingAt: offset)
+    let stride = _opaqueNextWordIndex(startingAt: start) &- start
+    _internalInvariant(offset <= start + stride, "Word breaking inconsistency")
+
+    if offset >= start + stride {
+      return i
+    }
+
+    return String._WordView.Index(_encodedOffset: start)
+  }
+
+  @inline(never)
+  @_effects(releasenone)
+  internal func _opaqueNextWordIndex(startingAt i: Int) -> Int {
```

Comment:
If this does not correspond to an opaque ABI entry point, why are we calling it `opaque`?

Nit: Given that `_WordView` is using this directly and this is already on an underscored type, I recommend removing the leading underscore here, but preserving it for the foreign path.